### PR TITLE
Fix bug in atomic_ref's calculation of lock_free-ness.

### PR DESCRIPTION
**Pull request by EricWF**
The builtin __atomic_always_lock_free takes into account the type of the
pointer provided as the second argument. Because we were passing void*,
rather than T*, the calculation failed. This meant that
atomic_ref<T>::is_always_lock_free was only true for char & bool.

This bug exists elsewhere in the atomic library (when using GCC, we fail
to pass a pointer at all, and we fail to correctly align the atomic like
_Atomic would).

This bug was not initially caught because we don't ever actually expect
a given value for `is_always_lock_free`. This problem is common
throughout atomic, where the tests have been written to assert that
_the value under test_ IS _the value under test_. Which leads to the
admission of bugs like this.

Further work is needed to clean up:

(A) Our detection of has-64-bit-atomics, which uses std::atomic to
determine if std::atomic is supported... (the type `LargeType` may be 64
bits in size, but it's required alignment is only 1 byte). This
configuration test was never intended to provide that information.

(B) The use of __atomic_is_always_lock_free in the GCC atomic
implementation, where we lie about whether a type is always lock free,
when the alignment for the std::atomic<T> is much smaller than required.
For example, struct Counter {int x; int y; };, which _Atomic Counter
aligns to 8 bytes, but our std::atomic<Counter> under GCC only aligns to
4, but still reports that the type is always lock free.

(C) std::atomic_ref<T>::required_alignment should often times be larger
than the natural alignment of the type if the sizeof(T) > alignof(T) and
sizeof(T) 2, 4, 8, or 16. (See the Counter example). In failing to do
so we make many types (Again, see Counter), non-lock free even when
there are atomic instructions on the host that support types of that
size.

(D) We need to actually test against hard coded values throughout our
atomic tests to avoid these sorts of bugs in the future. This probably
means auditing the entire atomic test suite.

This change attempts to start sorting out the testing difficulties
by using the __GCC_ATOMIC_(CHAR|SHORT|INT|LONG|LLONG|POINTER)_IS_LOCK_FREE
predefined macros to establish an expected value for
`is_always_lock_free` and `is_lock_free` for the respective types,
as well as types with matching sizes and compatible alignment values
(Where compatible alignment meants alignof(T) >=
alignof(char|short|int|long|long long) for the matching sized type).

Using these compiler pre-defines we can actually validate that certain
types, like char and int, are actually always lock free like they are
on every platform in the wild(*).

(*) At least for every platform we care about.

Fixing (B) reqires an ABI break where we bump the alignment on the
type std::atomic<T> to match that of _Atomic T (were we under clang).

Fixing (C) also requires an ABI break, but atomic_ref is new enough
that we should consider it ASAP. (Though fixing (C) is arguably more of
a QoI detail, but it's a big one, since we don't want the runtime
alignment of memory to determine the locking behavior of the atomic).

### CODE REVIEW

**Comment by huixie90** [comment]
I remember that we had a discussion on this in the original PR but I don’t remember the outcome. I was focusing on the clang bit. (IIRC My concern was that clang builtin had assumption that the address is always _Atomic)

https://github.com/llvm/llvm-project/pull/76647#discussion_r1514600368

File: libcxx/include/__atomic/atomic_ref.h [hunk 1]
```diff
@@ -42,6 +42,19 @@ _LIBCPP_BEGIN_NAMESPACE_STD
...
+// GCC won't allow the reinterpret_cast<_Tp*>(-required_alignment) trick initially used,
+// so we need to actually fake up an instance of a type with the correct alignment.
+template <size_t _Alignment>
+struct __alignment_checker_type {
+  alignas(_Alignment) char __data;
+};
+
+template <size_t _Alignment>
+struct __get_aligner_instance {
+  static constexpr __alignment_checker_type<_Alignment> __instance{};
```
Comment (dalg24) [inline libcxx/include/__atomic/atomic_ref.h:55]: Any reason for not just inlining the definition of the aligned static object?
```suggestion
  alignas(_Alignment) static constexpr char __instance{};
};
```

**Review by dalg24 — COMMENTED** [review]
Thanks, I like the general approach.
I just have a few minor comments.

**Comment by ldionne** [comment]
CC @jyknight

File: libcxx/test/std/atomics/atomics.ref/is_always_lock_free.pass.cpp [hunk 2]
```diff
@@ -18,9 +18,25 @@
...
+    static_assert(std::atomic_ref<T>::is_always_lock_free == (known_status == LockFreeStatus::always),
+                  "is_always_lock_free is inconsistent with known lock-free status");
+    if (known_status == LockFreeStatus::always) {
+      assert(a.is_lock_free() && "is_lock_free() is inconsistent with known lock-free status");
+    } else if (known_status == LockFreeStatus::never) {
+      assert(!a.is_lock_free() && "is_lock_free() is inconsistent with known lock-free status");
+    } else {
+      assert(a.is_lock_free() || !a.is_lock_free()); // This is kinda dumb, but we might as well call the function once.
+    }
+  }
```
[the remark this answers is not in the document]
Comment (EricWF) [reply libcxx/test/std/atomics/atomics.ref/is_always_lock_free.pass.cpp:39]: I'm not sure I fully understand what you're suggesting. 

I think we still want to check that a type that is always lock free is also lock free at runtime.

**Review by ldionne — COMMENTED** [review]
I like the general direction. What other tests should we fix while we're at it? It would probably make sense to tackle them in this review as well. I suspect there's not that many.

File: libcxx/test/std/atomics/atomics.ref/is_always_lock_free.pass.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply libcxx/test/std/atomics/atomics.ref/is_always_lock_free.pass.cpp:39]: > I think we still want to check that a type that is always lock free is also lock free at runtime.

You're right.

**Comment by aeubanks** [comment]
there's some fallout we're seeing internally from #98081 that would be fixed with this, any chance of this going through?

**Comment by ldionne** [comment]
I tried picking up this patch to apply some of the feedback but I couldn't push to @EricWF 's branch. Since this is starting to be tight for getting into LLVM 19, I created #99570 with this patch + some of my edits. Closing this PR since it's now a duplicate of the other one.

**Pull request closed without merging**